### ISSUE TITLE
[Question] Wasm / Node.js support

### ISSUE THREAD

**Issue by codetheweb** [librespot-org/librespot#530]
I'm new to Rust, and I was wondering how hard it would be to compile this into WebAssembly for usage with Node.js / browsers. I tried briefly using `wasm-pack build` (after adding `crate-type = ["cdylib"]` to Cargo.toml) and it just spit out a bunch of errors. I'm not experience enough to know whether this is relatively easy with a little bit more configuration or if it would require rewriting code.

Any feedback / help is appreciated.

**Comment by codetheweb** [comment]
After some more investigation, it seems that the main blocker on this is the fact that raw TCP sockets are used (which are supported in Node.js but not the browser).

If you want to run it in a browser, it might be possible to use Emcripten which polyfills sockets to web sockets. You'd also then need something like [websockify](https://github.com/novnc/websockify) on a server.

In theory it should be possible to run on Node.js using network calls from [WASI](https://github.com/WebAssembly/WASI) but I'm not sure if WASM compilers work with that yet. This is a very rapidly changing space.

So TL;DR is that if you want Node.js support it looks like it might be easiest to just re-implement in JS.

**Comment by schickling** [comment]
Thanks for sharing your investigations @codetheweb. 

Looks like there's some ongoing work on bringing [socket support to Wasi](https://github.com/WebAssembly/WASI/pull/312).
